### PR TITLE
ntfy-sh: add runtime secret file support

### LINKED ISSUE #352461: nixos/ntfy-sh: Add options for secrets files
## Issue description

The ntfy server has a few secrets inside it's configuration that are set as string values, for example the `smtp-sender-pass` and the `web-push-private-key`. For my setup tho, and every other setup that uses either sops-nix or agenix, secrets are stored in encrypted files that are decrypted only at runtime, making it impossible to declaratively set them at build time. To work around this, some services have options to set a file path to a secret, which is then read and injected at runtime through the preStart script of the systemd service. An example of this is [forgejo on 24.05](https://github.com/NixOS/nixpkgs/blob/nixos-24.05/nixos/modules/services/misc/forgejo.nix#L551) with the database.passwordFile (they switched their method on unstable).

It would help me, and hopefully many others, greatly to see such options implemented for ntfy's secrets.
I could probably do this myself, but i'm not familiar with the nixpkgs contribution process yet and would probably need some assistance.

## Notify maintainers
@arjan-s @fpletz 

## Technical details
```
$ nix run nixpkgs#nix-info -- -m
 - system: `"x86_64-linux"`
 - host os: `Linux 6.11.5-zen1, NixOS, 24.11 (Vicuna), 24.11.20241025.18536bf`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.24.9`
 - nixpkgs: `/nix/store/wb6agba4kfsxpbnb5hzlq58vkjzvbsk6-source`
```

### PR DESCRIPTION
**Pull request by user12349872398470129740892375**
ADDED: smtpSenderPassFile and webPushPrivateKeyFile options
ADDED: systemd LoadCredential for secure runtime secret handling
ADDED: security hardening options for notification service
CHANGED: state directory handling

**Closes #352461** 

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
**Note:** I attempted to run `nixpkgs-review`, but my dual core laptop repeatedly froze during the process. If additional testing is required, I would appreciate assistance or confirmation from a reviewer with more capable hardware.
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/modules/services/misc/ntfy-sh.nix [hunk 1]
```diff
@@ -1,14 +1,16 @@
 { config, lib, pkgs, ... }:
+
 let
   cfg = config.services.ntfy-sh;
-
+  # ADDED: Proper YAML format handling
```
Comment (pluiedev) [inline nixos/modules/services/misc/ntfy-sh.nix:5]: There's no need for these comments

**Comment by pluiedev** [comment]
Welcome to Nixpkgs! Please squash your commits into just one 🥰

File: nixos/modules/services/misc/ntfy-sh.nix (hunk 1 above)
Comment (JacobMetz) [reply nixos/modules/services/misc/ntfy-sh.nix]: Fixed!

**Comment by JacobMetz** [comment]
Thank you!

File: nixos/modules/services/misc/ntfy-sh.nix [hunk 2]
```diff
@@ -53,7 +94,8 @@ in
       '';
 
       description = ''
-        Configuration for ntfy.sh, supported values are [here](https://ntfy.sh/docs/config/#config-options).
+        Configuration for ntfy.sh, supported values are documented at
+        https://ntfy.sh/docs/config/#config-options
```
Comment (pluiedev) [inline nixos/modules/services/misc/ntfy-sh.nix:98]: IIRC bare links should be surrounded by angle brackets
```suggestion
        Configuration for ntfy.sh, supported values are documented at
        <https://ntfy.sh/docs/config/#config-options>.
```

File: nixos/modules/services/misc/ntfy-sh.nix [hunk 3]
```diff
@@ -82,11 +123,23 @@ in
...
+          WEB_PUSH_PRIVATE_KEY = lib.mkIf (cfg.webPushPrivateKeyFile != null)
+            "$(cat $CREDENTIALS_DIRECTORY/webpush_key)";
+        };
+
         serviceConfig = {
           ExecStart = "${cfg.package}/bin/ntfy serve -c ${configuration}";
           User = cfg.user;
           StateDirectory = "ntfy-sh";
 
+          LoadCredential = lib.mkIf (cfg.smtpSenderPassFile != null || cfg.webPushPrivateKeyFile != null)
```
Comment (pluiedev) [inline nixos/modules/services/misc/ntfy-sh.nix:139]: After thinking about this, I think it still works if `LoadCredential `and `environment` are just empty, right? So there should be no need for a mkIf here

File: nixos/modules/services/misc/ntfy-sh.nix (hunk 2 above)
Comment (JacobMetz) [reply nixos/modules/services/misc/ntfy-sh.nix:98]: Fixed!

File: nixos/modules/services/misc/ntfy-sh.nix (hunk 3 above)
Comment (JacobMetz) [reply nixos/modules/services/misc/ntfy-sh.nix:139]: Fixed!

**Comment by pluiedev** [comment]
Looks like you mis-rebased there. Got a lot of unrelated commits in here

**Comment by JacobMetz** [comment]
> Looks like you mis-rebased there. Got a lot of unrelated commits in here

I think I fixed this. Sorry I am still getting the hang of git.

**Comment by JacobMetz** [comment]
> Looks like you mis-rebased there. Got a lot of unrelated commits in here

I am not sure why but I can't seem to fix it. Do you mind helping me out?

**Comment by pluiedev** [comment]
I think you just did a mass ping.

**Comment by JacobMetz** [comment]
Oh my god sorry. Will do.

**Pull request closed without merging**